### PR TITLE
[BugFix] Fix map_element bug when const map and variable key

### LINKED ISSUE #37594: BE crash when query with array_map
### Steps to reproduce the behavior (Required)
1. execute query:
```
with temp AS (
         SELECT
             map{
                 "1": "value1",
                 "2": "value2",
                 "3": "value3"
             } as map
     )
     select
         array_map(["1", "2", "3"], x -> temp.map[x])
     from
 temp;
```

### Expected behavior (Required)
Everything normal

### Real behavior (Required)
ERROR 1064 (HY000): Backend node not found. Check if any backend node is down.backend: [127.0.0.1 alive: true inBlacklist: true]
Be crash:
```
*** Aborted at 1703201813 (unix time) try "date -d @1703201813" if you are using GNU date ***
PC: @          0x69ca648 starrocks::MysqlRowBuffer::push_string()
*** SIGSEGV (@0x7f678272f000) received by PID 6728 (TID 0x7f6787738640) from PID 18446744071603154944; stack trace: ***
    @          0x7daff1a google::(anonymous namespace)::FailureSignalHandler()
    @     0x7f67eee3a520 (unknown)
    @          0x69ca648 starrocks::MysqlRowBuffer::push_string()
    @          0x3ed43dc starrocks::ArrayColumn::put_mysql_row_buffer()
    @          0x6819922 starrocks::MysqlResultWriter::process_chunk()
    @          0x672b838 starrocks::pipeline::ResultSinkOperator::push_chunk()
    @          0x5dc92df starrocks::pipeline::PipelineDriver::process()
    @          0x668278e starrocks::pipeline::GlobalDriverExecutor::_worker_thread()
    @          0x69fafcb starrocks::ThreadPool::dispatch_thread()
    @          0x69f530a starrocks::Thread::supervise_thread()
    @     0x7f67eee8cac3 (unknown)
    @     0x7f67eef1d814 clone
    @                0x0 (unknown)
```

### StarRocks version (Required)
```
StarRocks > select current_version();
+-------------------+
| current_version() |
+-------------------+
| 3.2.1-79ee91d     |
+-------------------+
1 row in set (0.18 sec)
```

### PR DESCRIPTION
**Pull request by Seaven**
Why I'm doing:

What I'm doing:
Oh, the reason for this bug: map_element expression handle error when map-column is a constant-column(only one rows) and the element key is a variable-column(may be many rows). In the case, lambda expression array_map(["1", "2", "3"], x -> temp.map[x]), array_map will construct a column with three rows as the map-key column for the map_element expression.
replay SQL:

```
CREATE TABLE `t1` (
  `v1` bigint(20) NULL COMMENT "",
  `v2` bigint(20) NULL COMMENT "",
  `v3` bigint(20) NULL COMMENT ""
) ENGINE=OLAP 
DUPLICATE KEY(`v1`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`v1`) BUCKETS 3 
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"enable_persistent_index" = "false",
"replicated_storage" = "false",
"compression" = "LZ4"
);
insert into t1 values (1, 1, 1), (2,2,2);
select map{"1": "a", "2": b}[v1] from t1;
```

Bug1: map_elment_expr handle const map and variable key error,
Bug1.1 : open stage use constant-column, but MapExpr result not constant column, and row number(1) is error, so use _const_col error, I think it's a risk point on all-expression (without function-call-expr)

Bug2: map_expr handle const map error, return a variable map with error rows
Bug3: map_column.get_datum will ignore null key
Bug4. some array_function hanlde constant-column(semi-type) is error

Fixes https://github.com/StarRocks/starrocks/issues/37594

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by imay — COMMENTED** [review]
Why not add the test case in the issue into our sql-test?

File: be/src/exprs/map_element_expr.cpp [hunk 1]
```diff
@@ -28,95 +31,62 @@ class MapElementExpr final : public Expr {
...
+            num_rows = chunk->num_rows();
         }
-        auto size = inputs[0]->size();
-        DCHECK(chunk == nullptr || chunk->num_rows() == size); // chunk is null in tests.
 
-        auto dest_size = size;
-        if (all_const) {
-            dest_size = 1;
+        if (map_col->only_null() || key_col->only_null()) {
+            return ColumnHelper::create_const_null_column(num_rows);
```
Comment (fzhedu) [inline be/src/exprs/map_element_expr.cpp:53]: key column could be only null
```
mysql> select map{1:3,3:3,null:3}[null];
+---------------------------+
| map{1:3,3:3,NULL:3}[NULL] |
+---------------------------+
|                         3 |
+---------------------------+
```

File: be/src/exprs/map_element_expr.cpp [hunk 2]
```diff
@@ -28,95 +32,62 @@ class MapElementExpr final : public Expr {
...
+            num_rows = chunk->num_rows();
         }
-        auto size = inputs[0]->size();
-        DCHECK(chunk == nullptr || chunk->num_rows() == size); // chunk is null in tests.
 
-        auto dest_size = size;
-        if (all_const) {
-            dest_size = 1;
+        if (map_col->only_null()) {
+            return ColumnHelper::create_const_null_column(num_rows);
```
Comment (fzhedu) [inline be/src/exprs/map_element_expr.cpp:54]: RETURN_IF_COLUMNS_ONLY_NULL()?

File: be/src/exprs/array_functions.cpp [hunk 3]
```diff
@@ -160,6 +160,11 @@ class ArrayRemoveImpl {
             return (*null_map)[idx] != 0;
         };
 
+        auto* targets_col = &targets;
+        if constexpr (ConstTarget) {
+            targets_col = down_cast<const ConstColumn*>(&targets)->data_column().get();
+        }
+
```
Comment (fzhedu) [inline be/src/exprs/array_functions.cpp:167]: ArrayContainsImpl also has the same problems

File: be/src/exprs/map_element_expr.cpp (hunk 2 above)
Comment (Seaven) [reply be/src/exprs/map_element_expr.cpp:54]: OK but not necessary?

File: be/src/exprs/array_functions.cpp (hunk 3 above)
Comment (Seaven) [reply be/src/exprs/array_functions.cpp:167]: yeah, I resolved it

**Review by fzhedu — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by Seaven** [comment]
https://github.com/Mergifyio backport branch-3.1 branch-3.2

**Pull request merged**